File: jar-common/src/test/java/com/beancounter/common/TestFx.java
package com.beancounter.common;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.contracts.FxRequest;
import com.beancounter.common.contracts.FxResponse;
import com.beancounter.common.model.Currency;
import com.beancounter.common.model.CurrencyPair;
import com.beancounter.common.model.FxPairResults;
import com.beancounter.common.model.FxRate;
import com.beancounter.common.utils.RateCalculator;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.common.annotations.VisibleForTesting;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import org.junit.jupiter.api.Test;

class TestFx {

  @Test
  @VisibleForTesting
  void is_RateRequestSerializing() throws Exception {
    CurrencyPair pair = CurrencyPair.builder().from("THIS").to("THAT").build();
    Collection<CurrencyPair> pairs = new ArrayList<>();
    pairs.add(pair);
    FxRequest fxRequest = FxRequest.builder().pairs(pairs).build();
    ObjectMapper mapper = new ObjectMapper();
    String json = mapper.writeValueAsString(fxRequest);
    FxRequest fromJson = mapper.readValue(json, FxRequest.class);
    assertThat(fromJson)
        .isEqualToComparingFieldByFieldRecursively(fxRequest);

  }

  @Test
  @VisibleForTesting
  void is_FxRequestPairsIgnoringDuplicates() {
    CurrencyPair pair = CurrencyPair.builder().from("THIS").to("THAT").build();
    FxRequest fxRequest = FxRequest.builder().build();
    fxRequest.add(pair);
    fxRequest.add(pair);
    assertThat(fxRequest.getPairs()).hasSize(1);
  }

  @Test
  @VisibleForTesting
  void is_FxResultSerializing() throws Exception {
    CurrencyPair nzdUsd = CurrencyPair.builder().from("NZD").to("USD").build();
    CurrencyPair usdUsd = CurrencyPair.builder().from("USD").to("USD").build();

    Collection<CurrencyPair> pairs = new ArrayList<>();
    pairs.add(nzdUsd);
    pairs.add(usdUsd);
    Map<String, FxRate> rateMap = new HashMap<>();
    rateMap.put("NZD", FxRate.builder()
        .to(Currency.builder().code("NZD").build())
        .from(Currency.builder().code("USD").build())
        .rate(BigDecimal.TEN).build());

    rateMap.put("USD", FxRate.builder()
        .to(Currency.builder().code("USD").build())
        .from(Currency.builder().code("USD").build())
        .rate(BigDecimal.ONE).build());

    Map<String, FxPairResults> rateResults = new RateCalculator()
        .compute("2019/08/27", pairs, rateMap);

    ObjectMapper objectMapper = new ObjectMapper();
    FxResponse fxResponse = FxResponse.builder().data(rateResults).build();
    String json = objectMapper.writeValueAsString(fxResponse);
    FxResponse fromJson = objectMapper.readValue(json, FxResponse.class);
    assertThat(fromJson.getData()).isNotNull();
    assertThat(fromJson)
        .isNotNull()
        .isEqualToComparingFieldByField(fxResponse);

  }
}

File: jar-common/src/test/java/com/beancounter/common/TestFx.kt
package com.beancounter.common

import com.beancounter.common.TestMarkets.Companion.NZD
import com.beancounter.common.TestMarkets.Companion.USD
import com.beancounter.common.contracts.FxRequest
import com.beancounter.common.contracts.FxResponse
import com.beancounter.common.model.FxRate
import com.beancounter.common.model.IsoCurrencyPair
import com.beancounter.common.utils.BcJson
import com.beancounter.common.utils.RateCalculator
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.math.BigDecimal

/**
 * FX Request response tests.
 */
internal class TestFx {
    private val objectMapper = BcJson().objectMapper
    @Test
    fun is_DefaultPropertiesSet() {
        assertThat(FxRequest("2020-10-01").pairs).isNotNull
    }

    @Test
    fun is_FxRequestSerializationAssumptions() {
        val fxRequest = FxRequest()
        fxRequest.addTradeBase(IsoCurrencyPair(USD.code, NZD.code))
        fxRequest.addTradePf(IsoCurrencyPair(USD.code, NZD.code))
        fxRequest.addTradeCash(IsoCurrencyPair(USD.code, NZD.code))
        val json = objectMapper.writeValueAsString(fxRequest)
        val fromJson = objectMapper.readValue(json, FxRequest::class.java)
        assertThat(fromJson)
            .hasAllNullFieldsOrPropertiesExcept("rateDate", "pairs")
            .isEqualTo(fxRequest)
        assertThat(fromJson.pairs).hasSize(1)
    }

    @Test
    @Throws(Exception::class)
    fun is_RateRequestSerializing() {
        val pair = IsoCurrencyPair("THIS", "THAT")
        val fxRequest = FxRequest("")
        fxRequest.addTradeBase(pair)
        fxRequest.addTradePf(pair)
        assertThat(fxRequest.pairs).size().isEqualTo(1)
        val json = objectMapper.writeValueAsString(fxRequest)
        val fromJson = objectMapper.readValue(json, FxRequest::class.java)
        assertThat(fromJson.pairs).hasSize(1)
        assertThat(fromJson.tradeBase).isNull()
        assertThat(fromJson.tradePf).isNull()
        assertThat(fromJson.tradeCash).isNull()
    }

    @Test
    fun is_FxRequestPairsIgnoringDuplicates() {
        val pair = IsoCurrencyPair("THIS", "THAT")
        val fxRequest = FxRequest("")
        fxRequest.add(pair)
        fxRequest.add(pair)
        assertThat(fxRequest.pairs).hasSize(1)
    }

    @Test
    @Throws(Exception::class)
    fun is_FxResultSerializing() {
        val nzdUsd = IsoCurrencyPair(NZD.code, USD.code)
        val usdUsd = IsoCurrencyPair(USD.code, USD.code)
        val pairs: MutableCollection<IsoCurrencyPair> = ArrayList()
        pairs.add(nzdUsd)
        pairs.add(usdUsd)
        val rateMap: MutableMap<String, FxRate> = HashMap()
        val nzd = NZD
        val usd = USD
        val nzdRate = FxRate(usd, nzd, BigDecimal.TEN, null)
        rateMap[NZD.code] = nzdRate
        val usdRate = FxRate(usd, usd, BigDecimal.ONE, null)

        rateMap[USD.code] = usdRate
        val fxPairResults = RateCalculator.compute("2019/08/27", pairs, rateMap)

        val fxResponse = FxResponse(fxPairResults)
        val json = objectMapper.writeValueAsString(fxResponse)
        val fromJson = objectMapper.readValue(json, FxResponse::class.java)

        assertThat(fromJson.data).isNotNull
        assertThat(fromJson.data.rates).hasSize(2).containsKeys(nzdUsd, usdUsd)
    }
}
